text/0000-rand-crate-redesign.md

```diff
@@ -0,0 +1,1063 @@
+Implementations should be careful to prevent this from happening by sampling
+extra values and/or mutating the state. It should also be clear that this method
+does not add entropy, therefore should not be used for cryptography.
+
+Other constructors should be discouraged; e.g. all current generators have a
+`new_unseeded` function; realistically no one should use this except certain
+tests, where `SeedableRng::from_seed(seed) -> Self` could be used instead.
+
+Alternatively, `new` could seed from `thread_rng` or similar, or not exist
+forcing users to use `from_rng` or the `SeedableRng` trait. However, I believe
+`new` should exist and seed from `OsRng`, since this makes the easiest way to
+create an RNG secure and well seeded.
+
+## Generators
+
+This section concerns implementations of `Rng`;
+[the API is discussed above](#generation-api).
+
+In no particular order, this section attempts to cover:
+
+*   which generators should be provided by this library
+*   the story for generators in other crates
+*   requirements of generators included in this library
+*   benchmarks, testing and documentation on included generators
+*   convenience new types and functions
+
+### PRNG algorithms
+
+Rand currently provides three PRNG algorithms, and a wrapper (`OsRng`) to
+OS-provided numbers. It also provides a couple of simple wrappers, and two
+traits, [`Rng` and `SeedableRng`](#generation-api). The included PRNGs are:
+
+*   [`IsaacRng`] (32-bit) and [`Isaac64Rng`]; a very fast cryptographic
+    generator, but with potential attacks on weak states
+*   [`ChaChaRng`]; a cryptographic generator used by Linux and by Google for TLS,
+    among other uses
+*   [`XorShiftRng`]; a very fast generator, generally inferior to Xoroshiro
+
+**Question:** should any of the above be removed? What other PRNGs should we
+consider adding — should we keep the crate minimal or add several good
+generators? (Note that the questions of whether `rand` should be split into
+multiple crates and whether a separate crypto trait should be added affects
+this question.) [bhickey has some thoughts on this.](https://github.com/rust-lang-nursery/rand/pull/161#issuecomment-320483055).
+
+*   Likely [`XorShiftRng`] should be removed, since `Xoroshiro` is generally superior.
+*   Should we add [`Xoroshiro128+`] as a replacement for XorShift?
+    ([Wikipedia article](https://en.wikipedia.org/wiki/Xoroshiro128%2B))
+*   Should we add implementations for other promising crypto-PRNGs, e.g.
+    Fortuna/Yarrow (apparently used by *BSD, OSX, and iOS)?
+*   Should we add an implementation of [`RDRAND`]? This is supposed to be
+    secure and fast, but not everyone trusts closed-source hardware, and it may
+    not be the fastest. If we do, how should we handle platforms without this
+    feature?
+*   Wikipedia [mentions an improved `ISAAC+` variant](https://en.wikipedia.org/wiki/ISAAC_(cipher)#Cryptanalysis) of ISAAC; what is this?
```

Comment:
The ISAAC+ variant is described in the same [paper](states by https://eprint.iacr.org/2006/438.pdf) that mentions potential week states by Jean-Philippe Aumasson. It seems like a good improvement.

Reply:
On the subject of particular RNG types, the more I read of the PCG blog, it seems like Xoroshiro and friends aren't as great as they're first presented to be. They're super fast, but it seems to come at a price in quality. Obviously the PCG author might not be the most neutral source, but even the Xoroshiro example source mentions that the lowest three bits are known to be of poor quality.